scapy/contrib/coap_socket.py

```diff
@@ -0,0 +1,1001 @@
+    >>>         # and wait for whatever resource to be available.
+    >>>         TimeoutScheduler.schedule(1, self.delayed_message)
+    >>>         self.delayed_tokens.append((token, sa_ll))
+    >>>         return CoAPSocket.empty_ack_params()
+    >>> # Doesn't matter if it starts with "/dummy" or "dummy",
+    >>> # but it is an error if it is in the end
+    >>> lst_resources = [DummyResource("dummy"), DelayedResource("/delayed")].
+    >>> with CoAPSocket("127.0.0.1", 5683, lst_resources=lst_resources) as coap_socket:
+    >>>     while True:
+    >>>         pkg = coap_socket.recv()
+    >>>         # You can handle the packages inside your resources,
+    >>>         # here will only be the "unhandled" ones.
+
+    :param ip: ip address to bind udp socket to.
+    :param port: port to bind udp socket to.
+    :param ack_timeout: the time, in ms, that we should wait for the acknowledgment
+                        after sending a request.
+    :param retries: amount of retransmissions before giving up on the request.
+    :param duplication_response_timeout: Timeout, in fractions of seconds, that we will
+                                        keep the response in case a response get lost.
+    :param lst_resources: optional, list of registered resources.
+    :param sock: optional, a socket instance to transmit,
+                 if None, a classic UDP socket will be open and bound to ip/port.
+    :param close_on_timeout: Will try to close the socket if the retries is exceeded
+    """
+
+    def __init__(self,
+                 ip="",  # type: str
+                 port=5683,  # type: int
+                 ack_timeout=500,  # type: int
+                 retries=3,  # type: int
+                 duplication_response_timeout=1.00,  # type: float
+                 lst_resources=None,  # type: Optional[None, list[CoAPResource]]
+                 sock=None,  # type: Optional[None, SuperSocket, any]
+                 close_on_timeout=False  # type: bool
+                 ):
+        self.impl = CoAPSocketImpl(ip, port, ack_timeout, retries,
+                                   duplication_response_timeout, lst_resources, sock,
+                                   close_on_timeout)
+
+        self.ins = cast(socket.socket, self.impl)
+        self.outs = cast(socket.socket, self.impl)
+        self.basecls = CoAP
+
+    def recv_raw(self, x=0xffff):
+        # type: (int) -> Tuple[Optional[Type[Packet]], Optional[bytes], Optional[float]]
+        if not self.closed:
+            tup = self.impl.recv()
+            if tup is not None:
+                return self.basecls, tup[0], float(tup[1])
+        return self.basecls, None, None
+
+    def recv(self, x=MTU, **kwargs):
+        # type: (int, **Any) -> Optional[Packet]
+        return super(CoAPSocket, self).recv(x, **kwargs)
+
+    def close(self):
+        # type: () -> None
+        if not self.closed:
+            self.impl.close()
+            self.closed = True
```

Comment:
It’s better to set closed =True immediately, and then perform the closing action

Reply:
Done